relays/parachains/src/parachains_loop.rs

```diff
@@ -0,0 +1,1020 @@
+			continue
+		}
+
+		// we have no active transaction and may need to update heads, but do we have something for
+		// update?
+		let best_finalized_relay_block = target_client
+			.best_finalized_source_block(&best_target_block)
+			.await
+			.map_err(|e| {
+				log::warn!(
+					target: "bridge",
+					"Failed to read best finalized {} block from {}: {:?}",
+					P::SourceChain::NAME,
+					P::TargetChain::NAME,
+					e,
+				);
+				FailedClient::Target
+			})?;
+		let heads_at_source = read_heads_at_source(
+			&source_client,
+			&best_finalized_relay_block,
+			&sync_params.parachains,
+		)
+		.await?;
+		let updated_ids = select_parachains_to_update::<P>(
+			heads_at_source,
+			heads_at_target,
+			best_finalized_relay_block,
+		);
+		let is_update_required = is_update_required(&sync_params, &updated_ids);
+
+		log::info!(
+			target: "bridge",
+			"Total {} parachains: {}. Up-to-date at {}: {}. Needs update at {}: {}.",
+			P::SourceChain::NAME,
+			sync_params.parachains.len(),
+			P::TargetChain::NAME,
+			sync_params.parachains.len() - updated_ids.len(),
+			P::TargetChain::NAME,
+			updated_ids.len(),
+		);
+
+		if is_update_required {
+			let heads_proofs = source_client
+				.prove_parachain_heads(best_finalized_relay_block, &updated_ids)
+				.await
+				.map_err(|e| {
+					log::warn!(
+						target: "bridge",
+						"Failed to prove {} parachain heads: {:?}",
+						P::SourceChain::NAME,
+						e,
+					);
+					FailedClient::Source
+				})?;
+			log::info!(
+				target: "bridge",
+				"Submitting {} parachain heads update transaction to {}",
+				P::SourceChain::NAME,
+				P::TargetChain::NAME,
+			);
+			target_client
+				.submit_parachain_heads_proof(
+					best_finalized_relay_block,
+					updated_ids.clone(),
+					heads_proofs,
+				)
+				.await
+				.map_err(|e| {
+					log::warn!(
+						target: "bridge",
+						"Failed to submit {} parachain heads proof to {}: {:?}",
+						P::SourceChain::NAME,
+						P::TargetChain::NAME,
+						e,
+					);
+					FailedClient::Target
+				})?;
+
+			tx_tracker = Some(TransactionTracker::<P>::new(
+				updated_ids,
+				best_finalized_relay_block.0,
+				sync_params.stall_timeout,
+			));
+		}
+	}
+}
+
+/// Given heads at source and target clients, returns set of heads that are out of sync.
+fn select_parachains_to_update<P: ParachainsPipeline>(
+	heads_at_source: BTreeMap<ParaId, Option<ParaHash>>,
+	heads_at_target: BTreeMap<ParaId, Option<BestParaHeadHash>>,
+	best_finalized_relay_block: HeaderIdOf<P::SourceChain>,
+) -> Vec<ParaId>
+where
+	P::SourceChain: Chain<BlockNumber = RelayBlockNumber>,
+{
+	heads_at_source
+		.into_iter()
+		.zip(heads_at_target.into_iter())
+		.filter(|((para, head_at_source), (_, head_at_target))| {
+			let needs_update = match (head_at_source, head_at_target) {
+				(Some(head_at_source), Some(head_at_target))
+					if head_at_target.at_relay_block_number < best_finalized_relay_block.0 &&
+						head_at_target.head_hash != *head_at_source =>
```

Comment:
The idea (that still may be criticized, though) was to avoid decoding parachain heads in the bridge-prachains pallet code. That's because of two things:
1) polkadot is also not doing that in their parachain-related pallets - heads are always opaque there;
2) the pallet supports syncing multiple parachains at once, so if e.g. two parachains are using different header structs (i.e. one is frame-based and another is not), it'd require some non-trivial configuration.

But we still want to import parachain heads in order. To do that I'm relying on the fact that the best-finalized-parachain-head is always increasing. I.e. best finalized parachain head at relay block 100 is always the ancsestor of the best finalized parachain head at relay block 200. This also means that if we have already delivered best finalized parachain head at relay block 100, we don't need to do it again - it'll be the same.

This condition is exactly for handling this latter case: `head_at_target.at_relay_block_number < best_finalized_relay_block.0` means "if we MAY know better finalized-parachain-head than is currently available at the target" and the `head_at_target.head_hash != *head_at_source` means "and it is ACTUALLY BETTER". I.e. if first condition is true, and the second one is false, then the relay chain hasn't been updating parachain head for a while (it is a normal situation).

If we would change `head_at_target.at_relay_block_number < best_finalized_relay_block.0` to `head_at_target.at_relay_block_number <= best_finalized_relay_block.0`, then it'd mean "if we MAY know better OR THE SAME finalized-parachain-head than is currently available at the target". And we don't want to deliver the same parachain head multiple times (pallet would reject such transaction). But actually, it is safe to change this condition to "<=", because then the second condition would fail anyway and the parachain head won't pass filter. But imo "<" is stricter and safer to use here :)

Reply:
ok, that makes sense, let's keep the stricter/safer version and revisit if we see any need for lag improvements :+1: